keymanapp/keyman
# change(web): drops old IE-related handling code

Fixes #5943.
Also addresses some concerns of #3796.  As of commit https://github.com/keymanapp/keyman/pull/6557/commits/474634a90ed9657069ddf8b4518d863b89d649f5,

![image](https://user-images.githubusercontent.com/25213402/173538695-71ac3490-2f05-42ac-ae6e-b953cb31ebd9.png)


While this _should_ be safe, removing numerous old internal hooks like this could have unforeseen complications.

These changes are 16.0-only; they will not land in 15.0.

This PR seeks to remove a lot of the older parts of KMW that were devoted to IE-wrangling.  We haven't actively tested KMW on IE in a while and we officially dropped support versions ago.  Now, it's time to drop all pretenses of IE support entirely.

That said, we still do need to keep some polyfills around, as older Android devices that we still support launched with old enough versions of Chrome that parts of KMW will instantly break on them unless their users actually update Chrome.  I added documenting comments to the most relevant file:

https://github.com/keymanapp/keyman/blob/326dc4d72054dc51b16b67029a87029b243c4676/common/web/lm-worker/build.sh#L54-L80

This is the LMLayer's worker polyfill set.  Note that some of these polyfills are used explicitly because we don't want `es6-shim` bloat in the `WebWorker` - KeymanWeb itself would use `es6-shim` instead for such cases.  That said, KMW doesn't actually seem to utilize the features being polyfilled... here.

Also, now that we're dropping IE... we can actually just straight-up stop bundling `es6-shim` into KMW.

# User Testing 

## Test Suite 1:  in-app uses

SUITE_IN_APP:

### Testing Environments

GROUP_ANDROID_5_APP:  Android 5.0 / API 21 (any matching device; emulation is fine)
- Do NOT update Chrome if requested.
- If not requested and in an emulated environment, please reset the device to its "original settings" if possible.

GROUP_IOS_12_APP: iOS 12.0 (any matching device; Simulator is fine)

For both, please:
- Have `sil_euro_latin` (the default keyboard) installed with predictive text active.
- Have any other keyboard installed for at least one language.

### Tests

- **TEST_INAPP_BASE**:  Using English(EuroLatin SIL)...
  In portrait orientation, verify OSK is visible and fills the width the bottom of the screen

  1. Verify long-press `q` key works
  2. Verify long-press `p` key works
  3. Verify uppercase layer can be selected via `SHIFT`
  4. Verify number layer can be selected via `123`
  5. Verify long-press `1` key works
  6. Verify long-press `0` key works
  7. Verify backspace, space, and enter keys work
  8. Verify *Khmer* -> *Khmer Angkor* keyboard can be added via Settings menu

- **TEST_INAPP_LONG_PRESS**:  Using English(EuroLatin SIL)...

  Type on the OSK using the following scenarios and verify expected output:
  1. Clicking a suggestion on the suggestion banner - should insert the suggestion
  2. short-press a key and release - should insert the base key
  3. long-press a key, select a long-press key, and release - should insert the long-press key

- **TEST_KEYBOARD_SWAP**:  Use the keyboard's globe key to swap keyboards.
    - If any "Fatal Error" appears or the selected keyboard fails to load properly, _**FAIL**_ this test.

- **TEST_PREDICTIVE_TEXT**:  Using English(EuroLatin SIL)...
    1.  Use the app's "Clear Text" command to erase all text.
    2. The usual default suggestions should appear:  "the", "to", "a".
    3. Type `th`.  All suggestions should begin with these two letters.
    4. Select one of the suggestions.  It should replace the `th` with your selection.

## Test Suite 2:  standalone use on mobile platforms

SUITE_BROWSER:

### Testing Environments

GROUP_ANDROID_6_CHROME:  Chrome on Android **6.0 / API 23** (any matching device; emulation is fine)
- Note:  _**not**_ 5.0.  We do extra things in the app to support it that are no longer done for standalone-Web.
    - In particular, note the comments below regarding `for ... of` and `Map`.
GROUP_IOS_12_SAFARI: Safari on iOS 12.0 (any matching device; Simulator is fine)

For both, please:
- Have `sil_euro_latin` (the default keyboard) installed with predictive text active.
- Have any other keyboard installed for at least one language.

### Tests

- **TEST_INAPP_BASE**:  Use the testing page "Test unminified Keymanweb".
  In portrait orientation, verify OSK is visible and fills the width the bottom of the screen

  1. Verify long-press `q` key works
  2. Verify long-press `p` key works
  3. Verify uppercase layer can be selected via `SHIFT`
  4. Verify number layer can be selected via `123`
  5. Verify long-press `1` key works
  6. Verify long-press `0` key works
  7. Verify backspace, space, and enter keys work
  8. Verify *Khmer* -> *Khmer Angkor* keyboard can be added via Settings menu

- **TEST_LOREM_IPSUM**:  Use the testing page "Tests performance of touch-alias elements with large amounts of text" - it should be the last one listed on the testing index page.

    - Touch and drag the block of text, then tap somewhere within the text.  The caret should be placed near the location of the tap after a very small delay.  A bit of lagginess is fine, but it should still feel reasonably responsive.
    - Repeat this at least three times.
    - If the delay ever goes longer than one second, **_FAIL_** this test.
    - If the caret is not placed properly at any point, **_FAIL_**  this test.
    - After the final caret placement, type `lorem ipsum`.  Text entry should feel responsive and no laggier than the other interactions required by this test.
